RFC: FIR Merge - PR2: Add Fortran IR (FIR) MLIR dialect implementation

jeanPerier opened the pull request:
Add the FIR library that defines and implements the MLIR dialect used in flang.  FIR is defined and documented inside FIROps.td in this commit.

The following documents can help to understand how MLIR dialects must be defined:
- [MLIR LangRef](https://github.com/llvm/llvm-project/blob/master/mlir/docs/LangRef.md)  for general information about MLIR.
- [MLIR Table-driven Operation Definition Specification](https://github.com/llvm/llvm-project/blob/master/mlir/docs/OpDefinitions.md) can help you go through `FIROps.td`, `FIROpsSupport.h`, and `FIROps.h`.
- [MLIR Attributes and Types Definition](https://github.com/llvm/llvm-project/blob/master/mlir/docs/DefiningAttributesAndTypes.md) can help you go through `FIRAttr.h` and `FIRType.h`.

This patch adds the tool `tco` that processes FIR input files and drives transformations passes. The tco tool is used for testing. In this patch, the tco tool is to implement tests for parsing, printing, verifying, and round-tripping FIR.

Note:
This commit it a feature-based split of the work done in the FIR experimental branch. The related work log can be found in the commits between https://github.com/schweitzpgi/f18/commit/742edde572bd74d77cf7d447132ccf0949187fce and https://github.com/schweitzpgi/f18/commit/2ff55242126d86061f4fed9ef7b59d3636b5fd0b.

RFC: http://lists.llvm.org/pipermail/flang-dev/2020-January/000162.html

River707 commented on include/flang/Optimizer/Dialect/FIROps.td:
```diff
@@ -0,0 +1,2779 @@
+    if (resultType)
+      result.addTypes(resultType);
+    result.addOperands(operands);
+    for (auto namedAttr : attributes)
```
I'm pretty sure you can just do `result.addAttributes(attributes);`

River707 commented on include/flang/Optimizer/Dialect/FIROps.td:
```diff
@@ -0,0 +1,2779 @@
+    return mlir::success();
+  }];
+
+  let parser = [{
```
Use assemblyFormat?

River707 commented on include/flang/Optimizer/Dialect/FIROps.td:
```diff
@@ -0,0 +1,2779 @@
+  let builders = [OpBuilder<
+    "Builder *, OperationState &result, Value selector,"
+    "ArrayRef<Value> properOperands, ArrayRef<Block *> destinations,"
+    "ArrayRef<ArrayRef<Value>> operands = {},"
```
Generally `ValueRange` is a better parameter than `ArrayRef<Value>` to use for builders, as it can be constructed from operand/result/block/etc. ranges without constructing a vector.

River707 commented on include/flang/Optimizer/Dialect/FIROps.td:
```diff
@@ -0,0 +1,2779 @@
+        result.addSuccessor(std::get<0>(kvp), std::get<1>(kvp));
+      }
+      for (auto namedAttr : attributes) {
+        result.addAttribute(namedAttr.first, namedAttr.second);
```
Same as above, `result.addAttributes(attributes)` ?

River707 commented on include/flang/Optimizer/Dialect/FIROps.td:
```diff
@@ -0,0 +1,2779 @@
+    Arguments<(ins TypeAttr:$in_type, Variadic<AnyIntegerType>:$args)> {
+
+  let parser = [{
+    mlir::Type intype;
```
Could you please move all of these large code blocks to a .cpp source file? It is much easier to reason about/format/refactor code when it is in a .cpp file. It also makes the .td file much cleaner.

schweitzpgi replied:
Right. I'm wondering if such a NFC might be rolled up in another merge though to reduce any noise. (This is already a large PR.)

sscalpone replied:
@jeanPerier If this request is being deferred for now, please add it to the list of tasks for the FIR integration.

schweitzpgi replied:
We'll move some code around in a subsequent PR.

jeanPerier replied:
Added on list of task.

River707 commented on include/flang/Optimizer/Dialect/FIROps.td:
```diff
@@ -0,0 +1,2779 @@
+    Unlike LLVM's GEP instruction, one cannot stride over the outermost
+    reference; therefore, the leading 0 index must be omitted.
+
+      %i = ... : index
```
If you use ```mlir code blocks, you get syntax highlighting.

River707 commented on include/flang/Optimizer/Dialect/FIROps.td:
```diff
@@ -0,0 +1,2779 @@
+    // Recovering a LEN type parameter only makes sense from a boxed value
+    for (auto co : coor())
+      if (auto *s = co.getDefiningOp())
+        if (dyn_cast_or_null<LenParamIndexOp>(s)) {
```
You've already checked `s` for null, so you just need dyn_cast. Alternatively, you could merge the two if checks and continue using dyn_cast_or_null.

River707 commented on include/flang/Optimizer/Dialect/FIROps.td:
```diff
@@ -0,0 +1,2779 @@
+
+  let verifier = [{
+    auto size = triples().size();
+    if (size < 1 && size <= 16 * 3)
```
I don't quite parse this, it seems like `rhs` would always be true if `lhs` is true.

schweitzpgi replied:
Sleep deprivation. Fixed.

River707 commented on include/flang/Optimizer/Dialect/FIROps.td:
```diff
@@ -0,0 +1,2779 @@
+  let parser = "return parseLoopOp(parser, result);";
+
+  let printer = [{
+    p << getOperationName() << ' ' << getI
```
nit:  Optional parse methods are a bit weird. I've found that it makes it much easier to read if you wrap them with `succeeded`/`failed`.

River707 commented on include/flang/Optimizer/Dialect/FIROps.td:
```diff
@@ -0,0 +1,2779 @@
+  let parser = "return parseLoopOp(parser, result);";
+
+  let printer = [{
+    p << getOperationName() << ' ' << getI
```
Any reason not to have this attribute in the arguments list?

schweitzpgi replied:
Sorry. I'm not sure which attribute you're referring to here.

River707 replied:
`constant`. I don't see where it is added to an arguments list. There are a few other ops in a similar situation. Adding them to the argument list would also allow for using the assemblyFormat field.

schweitzpgi replied:
(Obvious now. Browser window goof?)

This one is tricky (and related to the conversation we were having on discourse). We want to be able to read in floating-point constants which don't correspond to the standard dialect supported semantics. I'll try mucking around with it a bit more.

jeanPerier replied:
Kept as it is in this PR. Added to the list of point to track.

River707 commented on include/flang/Optimizer/Dialect/FIROps.td:
```diff
@@ -0,0 +1,2779 @@
+  let parser = "return parseLoopOp(parser, result);";
+
+  let printer = [{
+    p << getOperationName() << ' ' << getI
```
Seems like this can use the `assemblyFormat` field.

schweitzpgi replied:
Does `assemblyFormat` support parsing Regions?

River707 replied:
Not yet, but it is on the list of stuff to add.

River707 commented on include/flang/Optimizer/Dialect/FIRType.h:
```diff
@@ -0,0 +1,357 @@
+  mlir::AffineMapAttr getLayoutMap() const;
+
+  static mlir::LogicalResult
+  verifyConstructionInvariants(const mlir::AttributeStorage *, mlir::Type eleTy,
```
The first param of these methods should be an `mlir::Location`.

schweitzpgi replied:
For some reason, a recent rebase to the tip of mono broke all these `verifyConstructionInvariants` functions. The parameter types were changed to sneak these past the compilers.

River707 replied:
Yeah, that was my fault. The original parameter was an `Optional<mlir::Location`, but I changed it to `mlir::Location`. A raw attribute storage is passed instead of a proper `mlir::Location` to avoid cyclic includes in the base storage uniquer class. I really need to add a static assert of some kind to ensure that clients still use Location.

schweitzpgi replied:
Ah. OK, let me try that...

River707 commented on lib/Optimizer/Dialect/FIRAttr.cpp:
```diff
@@ -0,0 +1,236 @@
+  if (attrName == ExactTypeAttr::getAttrName()) {
+    mlir::Type type;
+    if (parser.parseLess() || parser.parseType(type) || parser.parseGreater())
+      parser.emitError(loc, "expected a type");
```
Seems like you are missing some returns for these error cases. Lack of coverage?

schweitzpgi replied:
Fixed returns. We need to add negative tests.

River707 commented on lib/Optimizer/Dialect/FIRAttr.cpp:
```diff
@@ -0,0 +1,236 @@
+    return parseFirRealAttr(dialect, parser, type);
+
+  llvm::Twine msg{"unknown FIR attribute: "};
+  parser.emitError(loc, msg.concat(attrName));
```
```c++
parser.emitError("unknown FIR attribute: ") << attrName;
```

?

River707 commented on tools/tco/tco.cpp:
```diff
@@ -0,0 +1,119 @@
+
+void printModuleBody(mlir::ModuleOp mod, raw_ostream &output) {
+  // don't output the terminator bogo-op
+  auto e{--mod.end()};
```
Note: Block has a utility for this `Block::without_terminator`.

schweitzpgi replied:
Thanks. I was looking for that. :)

schweitzpgi commented:
> Hi Jean, I hope you don't mind but I left some comments. I still haven't been able to look through all of it though.

Thanks for the helpful tips, River. Much appreciated.

tskeith commented:
Can you update the build instructions? My LLVM build doesn't have AddMLIR.cmake but the install does. But the install doesn't have llvm-lit.

tskeith commented:
The tests fail for me with: `tco: command not found`

River707 commented on test-lit/Fir/fir-ops.fir:
```diff
@@ -0,0 +1,458 @@
+
+// CHECK-LABEL: @instructions
+func @instructions() {
+  // CHECK: %[[A0:.*]] = fir.alloca !fir.array<10xi32>
```
Not sure if you all are aware, but there is a script to help generate CHECK lines:
https://github.com/llvm/llvm-project/blob/master/mlir/utils/generate-test-checks.py

schweitzpgi replied:
Thanks. I'm making a note of this.

River707 commented on include/flang/Optimizer/Support/KindMapping.h:
```diff
@@ -0,0 +1,89 @@
+  using KindTy = unsigned;
+  using Bitsize = unsigned;
+  using LLVMTypeID = llvm::Type::TypeID;
+  using MatchResult = llvm::Optional<bool>;
```
What is this supposed to be used for? Is this just for parsing? Would it be better to do something like `mlir::ParseResult`?

schweitzpgi replied:
Right. This code doesn't use much mlir support code. It looks like we could easily use `mlir::OptionalParseResult` here though.

River707 commented on include/flang/Optimizer/Support/KindMapping.h:
```diff
@@ -0,0 +1,89 @@
+  MatchResult parse(llvm::StringRef kindMap);
+
+  mlir::MLIRContext *context;
+  std::map<char, std::map<KindTy, Bitsize>> intMap;
```
Do these need to be sorted? Would it be better to use an unordered container instead?

schweitzpgi replied:
I believe fast lookup is the only property really sought.

River707 commented on include/flang/Optimizer/Dialect/FIRType.h:
```diff
@@ -0,0 +1,357 @@
+                               mlir::Type eleTy);
+};
+
+class LenType
```
Is there a reason why some of these types have comments and others don't?

schweitzpgi replied:
Thanks. I'll document these.

River707 commented on lib/Optimizer/Dialect/FIRType.cpp:
```diff
@@ -0,0 +1,1288 @@
+
+// Len
+
+LenType fir::LenType::get(mlir::MLIRContext *ctxt, KindTy) {
```
Why does this have a KindTy parameter if it is never used?

schweitzpgi replied:
As I recall, there was some issue I had with uniquing these singleton types. I used a dummy KindTy to work around the issue at the time. We should investigate if that problem was fixed.

jeanPerier replied:
It turned out it is not needed anymore. KindTy parameter was removed.

River707 commented on lib/Optimizer/Dialect/FIRType.cpp:
```diff
@@ -0,0 +1,1288 @@
+      eleTy.isa<PointerType>() || eleTy.isa<ReferenceType>() ||
+      eleTy.isa<TypeDescType>() || eleTy.isa<SequenceType>()) {
+    auto context = attributeStorage->getDialect().getContext();
+    mlir::emitError(mlir::UnknownLoc::get(context),
```
emitError returns a diagnostic, which converts to `mlir::failure` automatically, so you can just return that directly.

River707 commented on lib/Optimizer/Dialect/FIRType.cpp:
```diff
@@ -0,0 +1,1288 @@
+                     const SequenceType::Shape &sh_2) {
+  if (sh_1.size() != sh_2.size())
+    return false;
+  for (std::size_t i = 0, e{sh_1.size()}; i != e; ++i)
```
Why the explicit std::size_t? Also why not simply `e = sh_1.size()`?

River707 commented on lib/Optimizer/Dialect/FIRType.cpp:
```diff
@@ -0,0 +1,1288 @@
+  return Base::get(ctxt, FIR_LEN, 0);
+}
+
+// LOGICAL
```
Shameless plug, but I've found the following comment block style to make chunking files much more readable:
https://github.com/llvm/llvm-project/blob/c3595d1069277b4ab0df49d7139b6f1bbc94f21c/mlir/lib/IR/Block.cpp#L15

schweitzpgi replied:
Noting this for a later PR.

River707 commented on lib/Optimizer/Dialect/FIROps.cpp:
```diff
@@ -0,0 +1,573 @@
+  return modBuilder.create<mlir::FuncOp>(loc, name, type, attrs);
+}
+
+GlobalOp createGlobalOp(mlir::Location loc, mlir::ModuleOp module,
```
LLVM style is to explicitly qualify functions:
https://llvm.org/docs/CodingStandards.html#use-namespace-qualifiers-to-implement-previously-declared-functions

LLVM style is also to have `using namespace`, and to limit anonymous namespaces to just classes:
https://llvm.org/docs/CodingStandards.html#anonymous-namespaces

schweitzpgi replied:
Ongoing activity.

River707 commented on lib/Optimizer/Dialect/FIROps.cpp:
```diff
@@ -0,0 +1,573 @@
+
+void LoopOp::build(mlir::Builder *builder, mlir::OperationState &result,
+                   int64_t lowerBound, int64_t upperBound, int64_t step) {
+  assert(false && "not implemented");
```
Can we move this to a later patch when it is needed?

River707 commented on lib/Optimizer/Dialect/FIROps.cpp:
```diff
@@ -0,0 +1,573 @@
+}
+
+mlir::ParseResult LoadOp::getElementOf(mlir::Type &ele, mlir::Type ref) {
+  if (mlir::Type r = elementTypeOf(ref)) {
```
```
if ((ele = ...))
  return ...
return ...
```
?

River707 commented on lib/Optimizer/Dialect/FIROps.cpp:
```diff
@@ -0,0 +1,573 @@
+
+/// Get the element type of a reference like type; otherwise null
+mlir::Type elementTypeOf(mlir::Type ref) {
+  if (auto r = ref.dyn_cast_or_null<ReferenceType>())
```
You could use mlir::TypeSwitch to simplify some of this:
```
TypeSwitch<mlir::Type>(ref)
  .Case<HeapType, PointerType, ReferenceType>([](auto type) { return type.getEleType(); })
  .Default([](mlir::Type) { return {}; });
```

River707 commented on lib/Optimizer/Dialect/FIROps.cpp:
```diff
@@ -0,0 +1,573 @@
+
+void GlobalOp::appendInitialValue(mlir::Operation *op) {
+  auto &block = getBlock();
+  block.getOperations().insert(block.end(), op);
```
.getOperations().push_back(op)
?

River707 commented on lib/Optimizer/Dialect/FIROps.cpp:
```diff
@@ -0,0 +1,573 @@
+  else
+    p << op.getOperand(0);
+  p << '(';
+  p.printOperands(llvm::drop_begin(op.getOperands(), isDirect ? 0 : 1));
```
You can do `p << ... << op.getOperands().drop_begin(...)`

schweitzpgi replied:
Tried this, but the compiler asks if I meant `drop_back` instead of `drop_begin`.

River707 replied:
Ahh, sorry I forgot that the naming is different: drop_front:
https://github.com/llvm/llvm-project/blob/0b6abe428164c38c7dec6c539e454cfd671067a4/mlir/include/mlir/Support/STLExtras.h#L255

Most ranges in MLIR use the above as a base, so most of the ArrayRef functionality is available automatically.

River707 commented on lib/Optimizer/Dialect/FIROps.cpp:
```diff
@@ -0,0 +1,573 @@
+
+/// Get the result types packed in a tuple tuple
+mlir::Type BoxDimsOp::getTupleType() {
+  llvm::SmallVector<mlir::Type, 3> triple{
```
nit: std::array<mlir::Type, 3> ?

schweitzpgi replied:
Well, it is a triple. :)

River707 commented on lib/Optimizer/Dialect/FIROps.cpp:
```diff
@@ -0,0 +1,573 @@
+  // Fortran semantics: C852 an entity cannot be both ALLOCATABLE and POINTER
+  // 8.5.3 note 1 prohibits ALLOCATABLE procedures as well
+  // FIR semantics: one may not allocate a memory reference value
+  if (intype.dyn_cast<ReferenceType>() || intype.dyn_cast<HeapType>() ||
```
Please use isa if you don't need the result.

River707 commented on lib/Optimizer/Dialect/FIROps.cpp:
```diff
@@ -0,0 +1,573 @@
+/// Create a legal memory reference as return type
+mlir::Type AllocaOp::wrapResultType(mlir::Type intype) {
+  // FIR semantics: memory references to memory references are disallowed
+  if (intype.dyn_cast<ReferenceType>())
```
Same here.

River707 commented on lib/Optimizer/Dialect/FIROps.cpp:
```diff
@@ -0,0 +1,573 @@
+        return true;
+  } else if (auto rt = inType.dyn_cast<fir::RecordType>()) {
+    // don't recurse if we're already visiting this one
+    for (auto name : visited)
```
llvm::is_contained(...)?

River707 commented on lib/Optimizer/Dialect/FIRAttr.cpp:
```diff
@@ -0,0 +1,236 @@
+    a.getValue().bitcastToAPInt().toStringUnsigned(ss, 16);
+    os << ss << '>';
+  } else {
+    assert(false && "attribute pretty-printer is not implemented");
```
nit: llvm_unreachable?

jeanPerier commented:
> Can you update the build instructions? My LLVM build doesn't have AddMLIR.cmake but the install does. But the install doesn't have llvm-lit.

> The tests fail for me with: tco: command not found

Thanks for testing, I have updated the README and also changed the cmake files so that `tco` is always built when `LINK_WITH_FIR=On` is set (which is default).

DavidTruby commented on lib/Optimizer/Dialect/FIRType.cpp:
```diff
@@ -0,0 +1,1294 @@
+                                       mlir::Type eleTy) {
+    assert(eleTy && "element type is null");
+    auto *storage = allocator.allocate<BoxProcTypeStorage>();
+    return new (storage) BoxProcTypeStorage{eleTy};
```
Suggestion: 
`return std::allocator_traits<decltype(allocator)>::construct<BoxProcTypeStorage>(allocator, eleTy);`

jeanPerier replied:
Thanks for the suggestion, I find the two liner less intimidating.

jeanPerier commented on include/fir/.clang-format:
```diff
@@ -1,2 +0,0 @@
-BasedOnStyle: LLVM
-AlwaysBreakTemplateDeclarations: Yes
```
No, I had added all the FIR related .clang-format in the previous PR. But in the meantime, this directory was deleted and replaced by `include/flang/Optimizer/Dialect`.

jeanPerier commented on include/flang/Optimizer/Dialect/FIROps.h:
```diff
@@ -0,0 +1,76 @@
+                 CmpFPredicate predicate, mlir::Value lhs, mlir::Value rhs);
+void buildCmpCOp(mlir::Builder *builder, mlir::OperationState &result,
+                 CmpFPredicate predicate, mlir::Value lhs, mlir::Value rhs);
+unsigned getCaseArgumentOffset(ArrayRef<mlir::Attribute> cases, unsigned dest);
```
It came from `using namespace mlir;` This using directive is required because of the generated `FIROps.h.inc` that does not use llvm and mlir namespace qualifier. I have qualified the `ArrayRef` you have mentioned and lowered the using directive to only impact the `FIROps.h.inc`  include.

schweitzpgi replied:
MLIR code often has implicit using on the `llvm` namespace. It's pretty ubiquitous, in fact.

jeanPerier commented on include/flang/Optimizer/Dialect/FIROps.h:
```diff
@@ -0,0 +1,76 @@
+#include "mlir/IR/OpImplementation.h"
+#include "mlir/IR/SymbolTable.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/StringRef.h"
```
Yes, they are used in the generated `FIROps.h.inc` that is included here and does not handle dependencies on its own.

jeanPerier commented on include/flang/Optimizer/Dialect/FIROpsSupport.h:
```diff
@@ -0,0 +1,73 @@
+/// FuncOp is created, and that new FuncOp is returned.
+mlir::FuncOp createFuncOp(mlir::Location loc, mlir::ModuleOp module,
+                          llvm::StringRef name, mlir::FunctionType type,
+                          llvm::ArrayRef<mlir::NamedAttribute> attrs = {});
```
They are accessible in both `FIROPs.h` and `mlir/Dialect/StandardOps/IR/Ops.h`. `StringRef` and `ArrayRef` are very common in all FIR/MLIR sources and MLIR also never re-include these headers at every level (see for instance https://github.com/llvm/llvm-project/blob/69d757c0e8ffc5b49fda10df38e470a56d616ef4/mlir/include/mlir/Dialect/StandardOps/IR/Ops.h#L256 in the header I just mentioned.)

schweitzpgi replied:
Yes, MLIR seems to follow the LLVM coding standard with respect to redundantly including header files.

https://llvm.org/docs/CodingStandards.html#minimal-list-of-includes

> you can include them either directly or indirectly through another header file.

jeanPerier commented on include/flang/Optimizer/Dialect/FIRType.h:
```diff
@@ -0,0 +1,397 @@
+class CharacterType
+    : public mlir::Type::TypeBase<CharacterType, mlir::Type,
+                                  detail::CharacterTypeStorage>,
+      public IntrinsicTypeMixin<CharacterType, TypeKind::FIR_CHARACTER> {
```
No, the whole point `IntrinsicTypeMixin` is to share the implement of functions `kindof` and `getId`, these functions need to be public.

DavidTruby replied:
Yes, I see, I don't think private inheritance quite does what I thought it did!

schweitzpgi commented on include/flang/Optimizer/Dialect/FIRType.h:
```diff
@@ -0,0 +1,397 @@
+bool isaMemref(mlir::Type t);
+
+/// Is `t` a FIR dialect aggregate type?
+bool isanAggregate(mlir::Type t);
```
`isa_foo` is one of those odd cases in LLVM that does use underscores. e.g., `isa_or_null` and `isa_and_nonnull`.  It's not unique. There are some others.

That said, we can flip a coin and make these cases consistent here.

schweitzpgi commented on include/flang/Optimizer/Support/KindMapping.h:
```diff
@@ -0,0 +1,90 @@
+
+  mlir::MLIRContext *context;
+  std::unordered_map<char, std::unordered_map<KindTy, Bitsize>> intMap;
+  std::unordered_map<char, std::unordered_map<KindTy, LLVMTypeID>> floatMap;
```
Thanks, yes. It was already on my list to replace it again.

schweitzpgi commented on lib/Optimizer/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+# Sources generated by tablegen have unused parameters.
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unused-parameter")
```
We rebase to the tip of LLVM and MLIR very frequently, so a lot of header file changes come and go. We need to be able to tolerate those changes without breaking our builds (-Werror), and unused parameters warnings do crop up fairly often.

DavidTruby replied:
Oh yes I see, my idea wouldn't work for files that include files from FIR, is that right?

schweitzpgi replied:
We do want the header files in Optimizer/* to compile without warnings. But, as LLVM is more permissive and the Optimizer library uses a good bit of LLVM, we want to be relaxed by default much like LLVM itself.

schweitzpgi commented on include/flang/Optimizer/Dialect/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+set(LLVM_TARGET_DEFINITIONS FIROps.td)
+mlir_tablegen(FIROps.h.inc -gen-op-decls)
+mlir_tablegen(FIROps.cpp.inc -gen-op-defs)
+add_public_tablegen_target(FIROpsIncGen)
```
Thank you for the suggestion. It does not work in my environment.

DavidTruby replied:
Having looked at it it looks like that function assumes you are inside the MLIR repo so that would be why it doesn't work.

schweitzpgi replied:
Darn. Well, that would explain it.

kiranchandramohan commented on include/flang/Optimizer/Dialect/FIROps.td:
```diff
@@ -0,0 +1,2844 @@
+  let parser = "return parseLoopOp(parser, result);";
+
+  let printer = [{
+    p << getOperationName()
```
Isn't LLVM's view that it will not reassociate floating point without the reassoc attribute? Does MLIR have any difference in treatment?

schweitzpgi replied:
Yes. I believe that is the case.
Well, this pertains to FIR specifically. So we want to respect the Fortran rules of when and where we can reassociate explicitly.

kiranchandramohan commented on lib/Optimizer/Dialect/FIRType.cpp:
```diff
@@ -0,0 +1,1294 @@
+  return isa_fir_type(t) || isa_std_type(t);
+}
+
+bool isa_memref(mlir::Type t) {
```
Is this isa_fir_memref?

jeanPerier commented on lib/Optimizer/Dialect/FIROps.cpp:
```diff
@@ -0,0 +1,571 @@
+      return true;
+    for (auto ext : shape)
+      if (ext < 0)
+        return true;
```
Thanks for the suggestion, I find the original code quit straightforward and more readable here.

jeanPerier commented on lib/Optimizer/Dialect/FIROps.cpp:
```diff
@@ -0,0 +1,571 @@
+    return {};
+  assert(ivArg.getOwner() && "unlinked block argument");
+  auto *containingInst = ivArg.getOwner()->getParentOp();
+  return dyn_cast_or_null<fir::LoopOp>(containingInst);
```
`fir::LoopOp` class is not a pointer, it is a class, but dyn_cast_or_null<T> can take T* arguments and then return T*.
See: https://github.com/llvm/llvm-project/blob/1bd6123b781120c9190b9ba58b900cdcb718cdd1/llvm/include/llvm/Support/Casting.h#L361

DavidTruby replied:
That's why I'm confused, the return statement has type LoopOp* but the function has return type LoopOp. I can't make that compile in my head.

schweitzpgi replied:
The LoopOp is a specific wrapper that holds a pointer to the Operation. The cast is promoting the Operation to that wrapper or returning an empty wrapper.

DavidTruby replied:
Sorry to keep questioning this, I'm sure I must be misunderstanding something. Does LoopOp have a constructor that takes a pointer to another LoopOp as a parameter then?

jeanPerier replied:
Ok, sorry for the wrong answer earlier. Here is my understanding of why this works:

- `containingInst` is an `mlir::Operation*`. See: https://github.com/llvm/llvm-project/blob/fcfd4fb686b83a5c8c7ba2c3bd4992ff24e83870/mlir/include/mlir/IR/OpDefinition.h#L109-L111

- All MLIR operations are using the curiously recursive template pattern and inherit from the `mlir::Op` class. This class provides a conversion operator from `mlir::Operation*` pointers that is intended to do the magic you see according to the related comment:
See: https://github.com/llvm/llvm-project/blob/fcfd4fb686b83a5c8c7ba2c3bd4992ff24e83870/mlir/include/mlir/IR/OpDefinition.h#L1038

- There is also a ctor form nullptr:
https://github.com/llvm/llvm-project/blob/fcfd4fb686b83a5c8c7ba2c3bd4992ff24e83870/mlir/include/mlir/IR/OpDefinition.h#L1034

Under the hood, all operations that are handled as value really just are a wrapper around a pointer. Note that this pointer can be null, in which case the operation cast to the `false` boolean when checked in if and asserts. 

You can have a look into the generated `include/flang/Optimizer/Dialect/FIROps.h.inc` in your build directory to see how `LoopOp` is actually defined.

River707 replied:
Essentially yes. The `*Op` classes are effectively nice "views" into an `Operation`, they hold no state other than an `Operation*`(1). To enable casting an `Operation` to one of these `*Op` classes, we provide custom cast implementations for llvm::isa/cast/dyn_cast to pick up on(2).


[1]: https://mlir.llvm.org/docs/Tutorials/Toy/Ch-2/#op-vs-operation-using-mlir-operations
[2]: https://github.com/llvm/llvm-project/blob/133db44996a80db957360a0b727479d4a8d4d1a2/mlir/include/mlir/IR/Operation.h#L685

DavidTruby replied:
Ok got it, thanks! Sorry to have kept poking at this I just thought it looked odd that we were going from LoopOp* to LoopOp during the return of this function. I guess I'll get used to this pattern when working with mlir

schweitzpgi commented on lib/Optimizer/Support/KindMapping.cpp:
```diff
@@ -0,0 +1,240 @@
+  return parseSingleChar<','>(ptr);
+}
+
+static MatchResult parseInt(unsigned &result, const char *&ptr) {
```
It's an ongoing issue that we've written code months before the evolving MLIR code adds or exposes similar functionality.

jeanPerier replied:
Will not be solved in this PR. Added on list of task.

schweitzpgi replied:
I looked at this a little more. I don't think we will pursue this one at this point. The "parser" here has nothing to do with MLIR syntax but is parsing a string that encodes the mapping of an arbitrary Fortran front-end's kind values to LLVM supported types. For example, in Bob's Fortran front-end, the type `INTEGER(KIND=747)` might correspond to an integer with 51 bits, an `i51` in LLVM IR. This kind mapping micro-language allows different front-ends (or potentially users) to configure these mappings at their choosing. Conceptually, it's like a data layout string in LLVM.

schweitzpgi commented on lib/Optimizer/Support/KindMapping.cpp:
```diff
@@ -0,0 +1,240 @@
+    return llvm::APFloat::x87DoubleExtended();
+  case LLVMTypeID::FP128TyID:
+    return llvm::APFloat::IEEEquad();
+  case LLVMTypeID::PPC_FP128TyID:
```
Thanks. It should be in parseTypeID. It's not part of the defaults because the front-end does not support it.

schweitzpgi commented on include/flang/Optimizer/Dialect/FIROps.h:
```diff
@@ -0,0 +1,76 @@
+class FirEndOp;
+class LoopOp;
+
+enum class CmpFPredicate {
```
I'll give this a shot. I seem to vaguely recall it used to be buried somewhere and not fully exposed.

River707 replied:
Note that I just recently(yesterday) moved the definition to tablegen:
https://github.com/llvm/llvm-project/commit/c10896682d0bb457b9d77fdd753ed9e7e6806db1#diff-cc3306bb0c567bb00f8a4a43b3145858

schweitzpgi replied:
Thanks for the heads-up. I'll rebase and make sure this hangs together. :)

schweitzpgi commented on include/flang/Optimizer/Dialect/FIROps.td:
```diff
@@ -0,0 +1,2844 @@
+  let parser = "return parseLoopOp(parser, result);";
+
+  let printer = [{
+    p << getOperationName()
```
Yes. That's correct.

schweitzpgi commented on include/flang/Optimizer/Dialect/FIROps.td:
```diff
@@ -0,0 +1,2844 @@
+  let parser = "return parseLoopOp(parser, result);";
+
+  let printer = [{
+    p << getOperationName()
```
It's not documented.
There is a type (3 actually) for referencing memory.

kiranchandramohan replied:
OK. But will there will be verification for valid conversions and also for lowering to LLVM dialect, right?
I was specifically asking whether the fir references can be converted to standard memrefs. Is there currently an operation which does that?

schweitzpgi replied:
Yes, We will define the restrictions and codify them in the verifier.

The plan will be to utilize standard passes including lowering to standard dialect when it is appropriate. However, doing such conversions is not in the scope of this PR, no.

kiranchandramohan replied:
OK. I should have clarified, I was not asking for it as part of this patch. But this question was for my information.
I have more questions relating to what will be a suitable type for use along with the OpenMP operations. Will it be standard memrefs or fir references? But i guess i have to discuss this with you separately and not here.

schweitzpgi replied:
That would be great and I'm happy to discuss with you. (But, yes, I am trying to keep the PR focused. Thanks!)

jeanPerier replied:
Will not be solved in this PR. Added on list of task.

schweitzpgi commented on include/flang/Optimizer/Dialect/FIROps.td:
```diff
@@ -0,0 +1,2844 @@
+  let parser = "return parseLoopOp(parser, result);";
+
+  let printer = [{
+    p << getOperationName()
```
Noted.

jeanPerier replied:
Will not be solved in this PR. Added on list of task.

schweitzpgi commented on include/flang/Optimizer/Dialect/FIROps.td:
```diff
@@ -0,0 +1,2844 @@
+  let parser = "return parseLoopOp(parser, result);";
+
+  let printer = [{
+    p << getOperationName()
```
For the Fortran use case, no. The programmer may specify an ordering to avoid overflows, for example, This is meant to explicitly enforce the determinism specified by the Fortran programmer, while allowing aggressive optimizations in the default case.

schweitzpgi commented on test-lit/Fir/fir-ops.fir:
```diff
@@ -0,0 +1,458 @@
+// FIR array types
+// CHECK-LABEL: func @arr1() -> !fir.array<10xf32>
+// CHECK-LABEL: func @arr2() -> !fir.array<10x10xf32>
+// CHECK-LABEL: func @arr3() -> !fir.array<?xf32>
```
Yes. This started out as one big test. Needs some cleanup.

jeanPerier replied:
Thanks for spotting that, I cleaned that up.

jeanPerier commented on lib/Optimizer/Dialect/FIRType.cpp:
```diff
@@ -0,0 +1,1288 @@
+      eleTy.isa<BoxProcType>() || eleTy.isa<DimsType>() ||
+      eleTy.isa<FieldType>() || eleTy.isa<LenType>() || eleTy.isa<HeapType>() ||
+      eleTy.isa<PointerType>() || eleTy.isa<ReferenceType>() ||
+      eleTy.isa<TypeDescType>())
```
Good point, thanks. Done.

jeanPerier commented on include/flang/Optimizer/Dialect/FIROps.h:
```diff
@@ -0,0 +1,76 @@
+  UNO,
+  // Always true
+  AlwaysTrue,
+  // Number of predicates.
```
This enum was removed to use the one from StandardOps, so this comment is not there anymore.

jeanPerier commented on include/flang/Optimizer/Dialect/FIRType.h:
```diff
@@ -0,0 +1,397 @@
+bool isa_fir_or_std_type(mlir::Type t);
+
+/// Is `t` a FIR dialect type that implies a memory (de)reference?
+bool isaMemref(mlir::Type t);
```
Changed to `isa_ref_type` to avoid ambiguity with MLIR memref.

jeanPerier commented on lib/Optimizer/Dialect/FIRDialect.cpp:
```diff
@@ -0,0 +1,77 @@
+namespace {
+
+template <typename A>
+void selectBuild(mlir::OpBuilder *builder, mlir::OperationState *result,
```
Not used anymore, removed.

psteinfeld commented on README.md:
```diff
@@ -93,6 +93,55 @@ LLVM_INSTALL_UTILS=On
+    -DLLVM_INSTALL_UTILS=On \
+    -DCMAKE_INSTALL_PREFIX=../install
+make -j
+make install
```
I tried these instructions on my Windows laptop using Windows Subsystem for Linux with Ubuntu installed and using GNU 8.3.0 as my compiler.  The "make -j" command caused my laptop to crash.  Presumably, it ran out of memory.  "make" without the "-j" command worked.  I did not try using an option like "make -j 8".  I also was able to build using Ninja rather than make.  This seemed to work better.  To do this, I just added the option "-G Ninja" at the front of the cmake command.  Specifically, my first line was 
```
cmake -G Ninja ../f18-llvm-project/llvm -DCMAKE_BUILD_TYPE=Release \
```
The command to do the build is then just
```
ninja
```
To do the install, it's
```
ninja install
```
This seemed to work better than using make.  Of course, it requires that you have Ninja installed on your build machine.

jeanPerier replied:
Thanks for testing the instructions ! I have removed the `-j`, that was a bit wild indeed. The examples are given with make to be portable and understandable by a wide audience. It is also consistent with other example (e.g. in `Build F18`). 
People familiar will builds will know how to replace this with Ninja and to set-up there `-j` according to their machine.

jeanPerier commented:
> Why is the LangRef not part of this? If this is autogen then can you add a pointer to how to access that?
Good point, it was not part of this PR because it require `add_mlir_function` to generate it but as mentioned in an earlier comment of @DavidTruby we cannot use it (it seems to only works for dialect whose sources live inside MLIR sources). So, I have replicated the part to generate the `FIRLangRef.md` and updated the README.md

jeanPerier commented:
> Is there no representation for co-arrays in FIR now?

That's correct. We want to gain more experience with both FIR and co-arrays before deciding how they should be handled in FIR.

jeanPerier commented:
Thanks all for your reviews !
We should have addressed all your comments so far (either solved, rejected, or agreed to solve them in later PRs as improvement). Please do not hesitate replying to this if you feel some of your comments were forgotten. Otherwise you are welcome to approve this PR so we can move to the next patches !

jeanPerier commented:
> Thanks for addressing the review comments. LGTM.
> 
> It will great if you can,
> 
> 1. Update the commit message with details of how to generate the documentation.
> 2. Spend some time to change or replace the original FIR investigation documentation
>    https://github.com/flang-compiler/f18/blob/master/documentation/Investigating-FIR-as-an-MLIR-dialect.md
> 3. Convey in the commit message what exactly is there in this patch and what is not. For e.g.,
>    -> contains the definition, printing, parsing and verification of operations.
>    -> is the verification of all operations there in this patch?
>    -> Co-arrays are not there.

Thanks @kiranchandramohan , I have removed https://github.com/flang-compiler/f18/blob/master/documentation/Investigating-FIR-as-an-MLIR-dialect.md because it was an early-design document that had a purpose when we started looking at FIR but is now outdated, thanks for pointing to it. The info inside is replaced by FIRLangRef.md that can be generated and new documents that will come with the next patches.

When squashing and rebasing for the merge, I will change the commit to:

```
Add Fortran IR (FIR) MLIR dialect implementation

Adds FIR library that implements an MLIR dialect to which Fortran
parse-tree will be lowered to.

FIR is defined and documented inside FIROps.td added in this commit.
It is possible to generate a more readable description FIRLangRef.md
from FIROps.td following the related instructions added to the README.md
by this commit.

This patch adds FIR definition and implementation that allow parsing,
printing, and verifying FIR. FIR transformations and lowering to Standard
and LLVM dialects are not part of this patch. The FIR verifiers are verifying
the basic properties of FIR operations in order to provide a sufficient
frame for lowering. Verifiers for more advanced FIR properties can be added
as needed.

Coarrays are not covered by FIR defined in this patch.

This patch also adds tco tool that is meant to process FIR input files and
drives transformations on it. The tco tool is used for testing.
In this patch, it is only used to demonstrate parsing/verifying/
and dumping FIR with round-trip tests.



Note:
This commit does not reflect an actual work log, it is a feature-based split of the
changes done in the FIR experimental branch. The related work log can be found in the
commits between:

https://github.com/schweitzpgi/f18/commit/742edde572bd74d77cf7d447132ccf0949187fce
and
https://github.com/schweitzpgi/f18/commit/2ff55242126d86061f4fed9ef7b59d3636b5fd0b

Changes on top of these original commits were made during this patch review.
```

jeanPerier commented:
I have pushed an update to adapt with the latest changes in MLIR (since end-of last week, this PR was could not be build with LLVM head due to MLIR interface changes). The `f18` branch of `f18-llvm-project` LLVM fork mentioned in the `README` was updated accordingly, but the current head from llvm-project also works with FIR at the time of this comment.

The other commit I pushed addresses @River707 comment that suggested moving big code pieces from the .td to a .cpp file (the comment was deferred, but the need to rebase with latest LLVM head gave us time to to it).

jeanPerier commented:
I have squashed and rebased to prepare merging, if you still need the commit history for the comments, you can find it here: https://github.com/flang-compiler/f18/commits/d0a7dabd6941a6725436e1c37ccc6e2b9aa0d6fb

DavidTruby commented:
F18 doesn't build for me since this merge, I get the following error:

```
/home/davtru01/f18/include/flang/Optimizer/Dialect/FIROps.td:17:9: error: Could not find include file 'mlir/Interfaces/ControlFlowInterfaces.td'
include "mlir/Interfaces/ControlFlowInterfaces.td"
```

I have mlir installed with LLVM and it finds AddMLIR.cmake correctly.

psteinfeld commented:
Similar to @DavidTruby, after pulling in these change, I cannot build.  I was able to build successfully a couple of days ago before these changes were merged into master.  My build environment is using GNU 8.3.0 and my own built version of LLVM, from the latest sources as of about a week ago, which is version 11.0.0.

The build fails very quickly.  Here's a fragment of the log file, starting at the beginning:
```
[1/190] Building FIROps.h.inc...
FAILED: include/flang/Optimizer/Dialect/FIROps.h.inc 
cd /mnt/c/GitHub/f18/master/Release && /mnt/c/GitHub/f18/fir/install/bin/mlir-tblgen -gen-op-decls -I -I -I /mnt/c/GitHub/f18/master/include/flang/Optimizer/Dialect -I /mnt/c/GitHub/f18/fir/install/include /mnt/c/GitHub/f18/master/include/flang/Optimizer/Dialect/FIROps.td --write-if-changed -o include/flang/Optimizer/Dialect/FIROps.h.inc -d include/flang/Optimizer/Dialect/FIROps.h.inc.d
/mnt/c/GitHub/f18/master/include/flang/Optimizer/Dialect/FIROps.td:17:9: error: Could not find include file 'mlir/Interfaces/ControlFlowInterfaces.td'
include "mlir/Interfaces/ControlFlowInterfaces.td"
        ^
/mnt/c/GitHub/f18/master/include/flang/Optimizer/Dialect/FIROps.td:17:9: error: Unexpected input at top level
include "mlir/Interfaces/ControlFlowInterfaces.td"
        ^
[2/190] Building FIROps.cpp.inc...
FAILED: include/flang/Optimizer/Dialect/FIROps.cpp.inc 
cd /mnt/c/GitHub/f18/master/Release && /mnt/c/GitHub/f18/fir/install/bin/mlir-tblgen -gen-op-defs -I -I -I /mnt/c/GitHub/f18/master/include/flang/Optimizer/Dialect -I /mnt/c/GitHub/f18/fir/install/include /mnt/c/GitHub/f18/master/include/flang/Optimizer/Dialect/FIROps.td --write-if-changed -o include/flang/Optimizer/Dialect/FIROps.cpp.inc -d include/flang/Optimizer/Dialect/FIROps.cpp.inc.d
/mnt/c/GitHub/f18/master/include/flang/Optimizer/Dialect/FIROps.td:17:9: error: Could not find include file 'mlir/Interfaces/ControlFlowInterfaces.td'
include "mlir/Interfaces/ControlFlowInterfaces.td"
        ^
/mnt/c/GitHub/f18/master/include/flang/Optimizer/Dialect/FIROps.td:17:9: error: Unexpected input at top level
include "mlir/Interfaces/ControlFlowInterfaces.td"
        ^
[3/190] Building CXX object lib/Common/CMakeFiles/FortranCommon.dir/Fortran.cpp.o
   ...
```

DavidTruby commented:
@psteinfeld, @jeanPerier mentioned to me that MLIR's interface changed on Tuesday. So you need a build of LLVM from yesterday or today for this to work correctly. Once I built LLVM from scratch from a new git clone, it worked fine.

psteinfeld commented:
> 
> 
> @psteinfeld, @jeanPerier mentioned to me that MLIR's interface changed on Tuesday. So you need a build of LLVM from yesterday or today for this to work correctly. Once I built LLVM from scratch from a new git clone, it worked fine.

Thanks!  I'll give it a try.

psteinfeld commented:
> 
> 
> > @psteinfeld, @jeanPerier mentioned to me that MLIR's interface changed on Tuesday. So you need a build of LLVM from yesterday or today for this to work correctly. Once I built LLVM from scratch from a new git clone, it worked fine.
> 
> Thanks! I'll give it a try.

I cleaned out my local LLVM build, rebuilt the latest from scratch, and now everything works.  Thanks for your help!

Note that my first attempt was to simple grab the latest sources from GitHub, run cmake, and rebuild.  That wasn't good enough.  I had to delete the build area and do everything from scratch.

How frequently should I expect to need to rebuild LLVM?

DavidTruby commented:
Once the new cmake changes land and we are part of the llvm monorepo, the easiest way of managing things should be to build alongside llvm and we shouldn't have these issues. See #1045 for more info!

psteinfeld commented:
> 
> 
> Once the new cmake changes land and we are part of the llvm monorepo, the easiest way of managing things should be to build alongside llvm and we shouldn't have these issues. See #1045 for more info!

Thanks, David!  You're up past my bed time.
